tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -277,5 +293,216 @@ StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+  const auto is_one_to_one_X_to_Y_cast = [](const HloInstruction* instr,
+                                            PrimitiveType X,
+                                            PrimitiveType Y) -> bool {
+    return (instr->opcode() == HloOpcode::kConvert &&
+            instr->shape().element_type() == Y && instr->operand_count() == 1 &&
+            instr->operand(0)->user_count() == 1 &&
+            instr->operand(0)->shape().element_type() == X);
+  };
+
+  if (match.conv->operand_count() < 4) {
+    // Conv input #3 (zero based) is side_input, after x, w, and bias.
+    // Side input doesn't exist in this case.
+    RewriteForConvertOrClampImpl(match);
+  } else if (is_one_to_one_X_to_Y_cast(match.conv->operand(3), S8, F32)) {
+    // If side_input has a convert_float_to_int8, absorb it as well.
+    auto side_converter = match.conv->mutable_operand(3);
+    side_converter->ReplaceAllUsesWithDifferentShape(
+        side_converter->mutable_operand(0));
+    side_converter->parent()->RemoveInstructionAndUnusedOperands(
+        side_converter);
+
+    RewriteForConvertOrClampImpl(match);
+  }
+}
+
+// Fuse the clamp/convert pattern with the int8 convolution custom call
+// (either pure or fused) for int8 output
+StatusOr<bool> RunFuseClamp(HloModule* module) {
+  bool changed = false;
+  for (HloComputation* computation : module->MakeNonfusionComputations()) {
+    std::vector<ConvWithConvertOrClamp> matches;
+    int num_forward_convs = 0;
+    for (auto instr : computation->instructions()) {
+      auto match = FindConvWithClampAndConvertToInt8(instr);
+      if (match.has_value()) {
+        matches.push_back(*match);
+      }
+    }
+    for (const ConvWithConvertOrClamp& match : matches) {
+      RewriteForFinalOutput(match);
+      changed = true;
+    }
+
+    // Report error for any convolution still having int32 output.
+    // Although int32 output convolution will trigger other sanity check errors
+    // later, we want to give specific error message here.
+    for (auto instr : computation->instructions()) {
+      if (auto call = DynCast<HloCustomCallInstruction>(instr)) {
+        if ((call->custom_call_target() == kCudnnConvForwardCallTarget ||
+             call->custom_call_target() ==
+                 kCudnnConvBiasActivationForwardCallTarget) &&
+            call->shape().tuple_shapes(0).element_type() == xla::S32) {
+          return Unimplemented(
+              "Integer convolutions for CuDNN must have float or int8 output.  "
+              "Use convert to cast output to float or the following pattern to "
+              "int8: "
+              "clamp(broadcast(-128), conv(int8_x, int8_w, ...), "
+              "broadcast(127)).");
+        }
+      }
+    }
+  }
+  return changed;
+}
+
+absl::optional<ConvWithConvertOrClamp> FindConvWithConvertToFloat(
+    HloInstruction* instr) {
+  using match::Convert;
+  using match::GetTupleElement;
+  using match::Op;
+
+  // The pattern we want to match:
```

Comment:
Could this comment become a docstring?

Reply:
> Could this comment become a docstring?

Sure.  I am happy to format the comments properly.  Do you have any docstring style guidance? This comment block tries to match the one in FindConvWithRelu().  I can tell the empty line at the end and wonder anything else to fix.

Reply:
The guideline is https://google.github.io/styleguide/cppguide.html#Function_Comments

To me, just removing the empty lines and putting it on top of the function would be sufficient.

Reply:
Thank you.  It is done.